library(dplyr)
library(tidyr)
library(ggplot2)
library(cowplot)
library(wesanderson)
library(VariantAnnotation)

#List of vcf files to look at
path2_GATK_vcf <- c(mode="character", length=4)
path2_GATK_vcf[1] <- "path_2_false_positive_snps_with_GATK_refence_NA12878"
path2_GATK_vcf[2] <- "path_2_false_positive_indels_with_GATK_refence_NA12878"
path2_GATK_vcf[3] <- "path_2_false_positive_snps_with_extended_GATK++_refence_NA12878"
path2_GATK_vcf[4] <- "path_2_false_positive_indels_with_extended_GATK++_refence_NA12878"

svp <- ScanVcfParam(info=c("BS","Regions"),geno=c("GT","BD","BK","BI","BVT","BLT","QQ"))
vcfs <- sapply(path2_GATK_vcf, readVcf,"hg19",svp)

#Generate a dataframe from the input vcfs:
make_df <- function(vcf,ref,type){
  df <- data.frame(ref=ref,type=type,chrom=as.character(seqnames(rowRanges(vcf))))
  df$pos <- start(ranges(rowRanges(vcf)))
  df$REF <- as.character(rowRanges(vcf)$REF)
  alt <- CharacterList(rowRanges(vcf)$ALT)
  df$ALT <- sapply(alt, function(x) paste(x,collapse=","))
  Regions <- unlist(info(vcf)$Regions)
  df$Regions_A <- Regions[start(PartitioningByEnd(info(vcf)$Regions))]
  df$Regions_B <- Regions[end(PartitioningByEnd(info(vcf)$Regions))]
  df$TRUTH_GT <- geno(vcf)$GT[,c("TRUTH")]
  df$CrG_GT <- geno(vcf)$GT[,c("QUERY")]
  df$CrG_BD <- geno(vcf)$BD[,c("QUERY")]
  df$CrG_BK <- geno(vcf)$BK[,c("QUERY")]
  df$CrG_BI <- geno(vcf)$BI[,c("QUERY")]
  df$CrG_BVT <- geno(vcf)$BVT[,c("QUERY")]
  df$CrG_BLT <- geno(vcf)$BLT[,c("QUERY")]
  df$CrG_QQ <- geno(vcf)$QQ[,c("QUERY")]
  df[names(df)[-c(4,5,6,16)]] <- lapply(df[names(df)[-c(4,5,6,16)]],as.factor)
  df$row.ids <- 1:dim(df)[1]
  df$GiaB <- FALSE
  return(df)
}
refs <- c("hg19","hg19","hg19_pp","hg19_pp")
types <- c("snps","indels","snps","indels")
df <- mapply(make_df,vcf=vcfs,ref=refs,type=types, SIMPLIFY = F)

#Mark the matching chrom,pos rows between hg19 and hg19_pp extended
snps_overlap <- merge(df[[3]],df[[1]],by=c("chrom","pos"),all.x=T) 
df[[3]]$GiaB[snps_overlap$row.ids.x[!is.na(snps_overlap$ref.y)]] <- TRUE
df[[1]]$GiaB <- TRUE

indels_overlap <- merge(df[[4]],df[[2]],by=c("chrom","pos"),all.x=T) 
df[[4]]$GiaB[indels_overlap$row.ids.x[!is.na(indels_overlap$ref.y)]] <- TRUE
df[[2]]$GiaB <- TRUE

df <- do.call("rbind", df)
df$row.ids <- NULL

#Are there duplicates? 
df[duplicated(df[,c("ref","type","chrom","pos")]),] # Nope!

#Read in the read count csv generated by the ipython script
read_counts <- read.csv("output_file_generated_by_FP_read_support_counter.py")
read_counts$ref <- rep(c("hg19","hg19","hg19_pp","hg19_pp"),times=sapply(vcfs,dim)[1,])
read_counts$type <- "snps"
read_counts$type[grep("indels",read_counts$vcf.ID)] <- "indels"
read_counts$vcf.ID <- NULL
read_counts$X <- NULL

#Merge the two dfs
df <- merge(df, read_counts, by=c("ref","type","chrom","pos"), all.x=T)

#Criteria for correcting a FP:
# >2 alt counts and >25% allele fraction
df$PacBio_support <- ifelse(df$PacBio_COV < 10,"No evidence",ifelse(df$PacBio_ALT>2 & (df$PacBio_ALT/df$PacBio_COV)>0.25,"Yes","No"))
df$PacBio_support <- as.factor(df$PacBio_support)
df$PacBio_support <- factor(df$PacBio_support,levels(df$PacBio_support)[c(3,1,2)])

#Summary stats
df %>% group_by(ref,type,PacBio_support) %>% summarize(count=n())

#Summary stats exclusive to GIAB++
exclusive <- split(df,df$ref)[[2]]
exclusive <- exclusive[!(exclusive$GiaB),]
exclusive %>% group_by(type,PacBio_support) %>% summarize(count=n())

#Generate Supplementary Figure S5
barplot <- df[-which(df$ref=="hg19_pp" & df$GiaB=="TRUE"),] %>% group_by(ref,type,PacBio_support) %>% summarize(count=n())
barplot <- melt(as.data.frame(barplot),id=c("ref","type","PacBio_support"))
levels(barplot$ref) <- c("GIAB","exclusive to GIAB++")
black.10.45.text <- element_text(color = "black", size = 12, angle= 45, hjust = 1)
colors=wes_palette("Darjeeling1",2)

barplot2 <- ggplot(barplot, aes(fill=PacBio_support, y=value, x=type)) + 
  geom_bar(position="dodge", stat="identity") +
  facet_wrap(~ref) +
  theme(axis.text.x=black.10.45.text, axis.text.y=element_text(color="black",size=10)) +
  ylab("") +xlab("") +
  scale_fill_manual(name ="PacBio support",values=c("#00A08A","#FF0000","#4D4D4D"))

png("SuppFig5_FP_PBsupport.png",type="cairo",height = 7, width = 7, units = 'in', res=300,bg="transparent")
barplot2
dev.off()

#Write the hg19 FP support to a csv
df2write <- df %>% dplyr::filter(ref=="hg19_pp") %>% dplyr::select(-c(ref,Regions_A,Regions_B,CrG_BD,CrG_BK,CrG_BI,CrG_BVT,CrG_BLT,CrG_QQ,GiaB))
df2write <- df2write[,c(1,18,2:7,12:15,8:11,16:17)]
names(df2write)[2] <- "PacBio_validated"
write.table(df2write,"Supplemetal_file1",row.names = F,quote = F,sep=",")

